pkg/shared/utils/sync/distributed_lock.go

```diff
@@ -0,0 +1,54 @@
+package sync
+
+import (
+	"github.com/pkg/errors"
+	"gorm.io/gorm"
+)
+
+// DistributedLock manages locks on the database through the usage of the `distributed_lock` table.
+// WARNING: a lock can be released only by the same instance that created it.
+type DistributedLock interface {
+	Lock() error
+	Unlock() error
+}
+
+var _ DistributedLock = &distributedLock{}
+
+// NewDistributedLock returns a new instance of DistributedLock
+// db - the connection to the database
+// lockID - the ID of the lock. If two threads calls the Lock function on the same lockID, they will be synchronised.
+func NewDistributedLock(db *gorm.DB, lockID string) DistributedLock {
+	return &distributedLock{db: db, id: lockID}
+}
+
+type distributedLock struct {
+	db *gorm.DB
+	tx *gorm.DB
+	id string
+}
+
+// Lock create the lock. If this lock is already locked, waits for the lock to be released
+func (dl *distributedLock) Lock() error {
+	dl.tx = dl.db.Begin()
+	result := dl.tx.Exec("INSERT INTO distributed_locks (ID) VALUES (?) ON CONFLICT DO NOTHING", dl.id)
```

Comment:
What kind of lock is acquired with the insert into?
Have we checked that other transactions can read the table without blocking when looking for another ID 'y', at the same time that a transaction is currently locking ID 'x'?
Using the pg_lock table might help to figure out what kind of lock is explicitly being used.
I am aware that there's a unit test where the intention is to test that case but it does not really use the database so the real behavior is not being tested. For that an integration test might be more appropriate.

Reply:
There are both unit and integration tests. An integration test has already been added to test on a real database.

> Have we checked that other transactions can read the table without blocking when looking for another ID 'y', at the same time that a transaction is currently locking ID 'x'?

Yes. No read is ever done into the table. Only `INSERT` is done and is rollbacked when the lock is unlocked.
Only a write lock is created for the newly inserted row.

Reply:
Added a new object `DistributedLockMgr`. Integration tests are [here](https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pull/1687/files#diff-4ccf384f6c30940554c1750d115d0684af071e71dfec7578b8e7c1d25557a763).

This new utility allows to easily manage different locks with different lock IDs.